### PR TITLE
[NS-55] Packed Transfers after Exchange

### PR DESCRIPTION
**Pull request by q10**
Summary:

- Update `VeColBatch.toStream()` and `VeColBatch.fromStream()` to move data to the VE using packed transfers after VE exchange.  In the new serialization format, the `UnitColVector` representation of the column vector metadata is serialized, followed by the transfer buffer bytes.  Upon deserialization, the `UnitColVector` and buffer are deserialized from the stream and packed into a `UcvTransferDescriptor` which will be used to transfer the data to the VE.
- Comment out `ColumnarBatchToVeColBatch` and related tests, which are not used anymore
- De-duplicate `vectorAlignedSize()` into `TransferDescriptor.vectorAlignedSize()`
- Clean up `VeDeserializationStream` and `VeSerializationStream` classes
- Clean up `VeSerializer` and `VeSerializerInstance` to require less arguments for construction
- Update `VectorEngineimpl.executeTransfer()` to fetch library reference of LibCyclone if available - this is required for `VeDeserializationStream` to work correctly when SparkCyclone is run in YARN context
- Update `VeColBatchUnitSpec` with more thorough serialization and deserialization tests
- Remove commented-out files
- Add unit tests for `UcvTransferDescriptor` class
- Update `VeSerializerSpec` with more thorough test cases
- Adjust the `TransferDescriptor` buffer size to be exact size, instead of double-counting the header size offsets

### CODE REVIEW

File: src/main/scala/com/nec/cache/TransferDescriptor.scala [hunk 1]
```diff
@@ -22,3 +23,47 @@ trait TransferDescriptor {
...
+    if (dangling > 0) {
+      // If the size is not aligned on 8 bytes, add some padding
+      size + (8 - dangling)
+    } else {
+      size
+    }
+  }
+
+  def bufferSize(columns: Seq[UnitColVector]): Long = {
+    require(columns.nonEmpty, "Need more than 1 columns to compute buffer size!")
```
Comment (treo) [inline src/main/scala/com/nec/cache/TransferDescriptor.scala:39]: This message looks a bit wrong, did you mean to say "need more than 0" or "Need at least 1"?

File: src/main/scala/com/nec/cache/TransferDescriptor.scala [hunk 2]
```diff
@@ -22,3 +23,47 @@ trait TransferDescriptor {
...
+      }
+
+    val dataOffsets: Seq[Long] = {
+      columns.flatMap(_.bufferSizes)
+        // Get the size of each buffer in bytes
+        .map { bufsize => vectorAlignedSize(bufsize.toLong) }
+        // Start the accumulation from header total size
+        // Offsets are in bytes
+        .scanLeft(headerOffsets.last * 8)(_ + _)
+    }
```
Comment (treo) [inline src/main/scala/com/nec/cache/TransferDescriptor.scala:64]: We are repeating this piece of code in a few places. Any good ideas on pulling that out into its own helpers, too?

**Review by treo — APPROVED** [review]
A few nitpicks, that could be addressed, but looks good in general.

File: src/main/scala/com/nec/cache/TransferDescriptor.scala (hunk 1 above)
Comment (q10) [reply src/main/scala/com/nec/cache/TransferDescriptor.scala:39]: Ah yes, that's a misspelling

File: src/main/scala/com/nec/cache/TransferDescriptor.scala (hunk 2 above)
Comment (q10) [reply src/main/scala/com/nec/cache/TransferDescriptor.scala:64]: I'll work on this in a subsequent PR.

**Pull request merged**